chainer/dataset/convert.py

```diff
@@ -139,3 +145,150 @@ def _concat_arrays_with_padding(arrays, padding):
             result[(i,) + slices] = src
 
     return result
+
+
+class ConcatWithAsyncTransfer(object):
+
+    """Interface to concatenate data and transfer them to GPU asynchronously.
+
+    It enables to transfer next batch of input data to GPU while GPU is
+    running kernels for training using current batch of input data.
+
+    Args:
+        stream (cupy.cuda.Stream): CUDA stream. If ``None``, a stream is
+            automatically created on the first call. Data transfer operation
+            is launched acynchrnously using the stream.
+    """
+
+    def __init__(self, stream=None):
+        self.stream = stream
+        self.conveyer = {}
+
+    def __call__(self, batch, device=None, padding=None):
+        """Concatenate data and transfer them to GPU asynchronously.
+
+        (see concat_examples for detail about concatenation)
+
+        Args:
+            batch (list): A list of examples.
+            device (int): Device ID to which each array is sent.
+            padding: Scalar value for extra elements.
+
+        Returns:
+            Array, a tuple of arrays, or a dictionary of arrays.
+            The type depends on the type of each example in the batch.
+        """
+        if len(batch) == 0:
+            raise ValueError('batch is empty')
+
+        first_elem = batch[0]
+
+        if device is not None and device >= 0 and self.stream is None:
+            with cuda.get_device_from_id(device):
+                self.stream = cuda.Stream(non_blocking=True)
+
+        if isinstance(first_elem, tuple):
+            result = []
+            if not isinstance(padding, tuple):
+                padding = [padding] * len(first_elem)
+
+            for i in six.moves.range(len(first_elem)):
+                if i not in self.conveyer:
+                    self.conveyer[i] = Conveyer(device, self.stream)
+                a = _concat_arrays(
+                    [example[i] for example in batch], padding[i])
+                a = self.conveyer[i](a)
+                result.append(a)
+
+            for i in six.moves.range(len(first_elem)):
+                self.conveyer[i].synchronize()
+
+            return tuple(result)
+
+        elif isinstance(first_elem, dict):
+            result = {}
+            if not isinstance(padding, dict):
+                padding = {key: padding for key in first_elem}
+
+            for key in first_elem:
+                if key not in self.conveyer:
+                    self.conveyer[key] = Conveyer(device, self.stream)
+                a = _concat_arrays(
+                    [example[key] for example in batch], padding[key])
+                a = self.conveyer[key](a)
+                result[key] = a
+
+            for key in first_elem:
+                self.conveyer[key].synchronize()
+
+            return result
+
+        else:
+            return to_device(device, _concat_arrays(batch, padding))
+
+
+class Conveyer(object):
```

Comment:
Should probably be "Conveyor" (conveyer is a rare alternative spelling)

Reply:
Thank you for pointing it out! Will change to "Conveyor".